File: articles/genes-inheritance-principles-of-heredity/r/quantitative_trait_heritability.R
# Simple quantitative-genetics scaffold in R.
#
# Estimates additive and phenotypic variance, h2, and response to selection.

library(dplyr)

qt_path <- file.path("data", "quantitative_trait.csv")

if (!file.exists(qt_path)) {
  qt_path <- file.path("..", "data", "quantitative_trait.csv")
}

trait_df <- read.csv(qt_path)

VA <- var(trait_df$additive_genetic_value)
VP <- var(trait_df$phenotype)
h2 <- VA / VP

selection_threshold <- quantile(trait_df$phenotype, 0.80)
selected <- trait_df$phenotype >= selection_threshold

S <- mean(trait_df$phenotype[selected]) - mean(trait_df$phenotype)
R <- h2 * S

summary_df <- data.frame(
  additive_variance = VA,
  phenotypic_variance = VP,
  h2 = h2,
  selection_differential = S,
  predicted_response = R
)

print(round(summary_df, 4))

File: articles/genes-inheritance-principles-of-heredity/python/quantitative_trait_heritability.py
"""
Quantitative trait and heritability scaffold.

Run:
    python python/quantitative_trait_heritability.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
QT_PATH = ARTICLE_DIR / "data" / "quantitative_trait.csv"


def main() -> None:
    """Estimate simple variance components and selection response."""

    df = pd.read_csv(QT_PATH)

    VA = df["additive_genetic_value"].var(ddof=1)
    VP = df["phenotype"].var(ddof=1)
    h2 = VA / VP

    selection_threshold = df["phenotype"].quantile(0.80)
    selected = df["phenotype"] >= selection_threshold

    S = df.loc[selected, "phenotype"].mean() - df["phenotype"].mean()
    R = h2 * S

    summary = pd.DataFrame(
        {
            "additive_variance": [VA],
            "phenotypic_variance": [VP],
            "h2": [h2],
            "selection_differential": [S],
            "predicted_response": [R],
        }
    )

    print(summary.round(4).to_string(index=False))


if __name__ == "__main__":
    main()
